src/repeated_game.jl

```diff
@@ -0,0 +1,339 @@
+    # Need slack variable for every subgradient and additional 2 incentive constraint
+    nH = size(H, 1)
+    nslack = nH + 2
+
+    # Create the c vector (objective)
+    c = zeros(nslack+2)
+
+    # Create the A matrix (constraints)
+    A_H = hcat(H, eye(nH, nslack))
+    A_IC_1 = zeros(1, nslack+2)
+    A_IC_2 = zeros(1, nslack+2)
+    A_IC_1[1] = -rpd.δ
+    A_IC_1[end-1] = 1.0
+    A_IC_2[2] = -rpd.δ
+    A_IC_2[end] = 1.0
+    A = vcat(A_H, A_IC_1, A_IC_2)
+
+    # Create the b vector (constraints)
+    b = Array(Float64, nslack)
+
+    return c, A, b
+end
+
+"""
+Given a constraint w ∈ W, this finds the worst possible payoff for agent i
+
+The output of this function is used to create the values associated with
+incentive compatibility constraints
+"""
+function worst_value_i(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}, i::Int)
+    # Objective depends on which player we are minimizing
+    c = zeros(2)
+    c[i] = 1.0
+
+    # Lower and upper bounds for w
+    lb = [-Inf, -Inf]
+    ub = [Inf, Inf]
+
+    lpout = linprog(c, H, '<', C, lb, ub, ClpSolver())
+    if lpout.status == :Optimal
+        out = lpout.sol[i]
+    else
+        out = minimum(rpd.sg.players[i].payoff_array)
+    end
+
+    return out
+end
+
+worst_value_1(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}) =
+    worst_value_i(rpd, H, C, 1)
+worst_value_2(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}) =
+    worst_value_i(rpd, H, C, 2)
+worst_values(rpd::RepeatedGame, H::Array{Float64, 2}, C::Array{Float64, 1}) =
+    (worst_value_1(rpd, H, C), worst_value_2(rpd, H, C))
+
+#
+# Outer Hyper Plane Approximation
+#
+"""
+Approximates the set of equilibrium value set for a repeated game with the
+outer hyperplane approximation described by Judd, Yeltekin, Conklin 2002
+"""
+function outerapproximation(rpd::RepeatedGame; nH=32, tol=1e-8, maxiter=500, nskipprint=1)
+    # Long unpacking of stuff
+    sg, δ = unpack(rpd)
+    p1, p2 = sg.players
+    po_1, po_2 = p1.payoff_array, p2.payoff_array
+    p1_minpayoff, p1_maxpayoff = extrema(po_1)
+    p2_minpayoff, p2_maxpayoff = extrema(po_2)
+
+    # Get number of actions for each player
+    nA1, nA2 = num_actions(p1), num_actions(p2)
+    nAS = nA1 * nA2
+
+    # Create action space
+    AS = QuantEcon.gridmake(1:nA1, 1:nA2)
+
+    # Create the unit circle, points, and hyperplane levels
+    C, H, Z = create_unitcircle_points(rpd, nH)
+    Cnew = copy(C)
+
+    # Create matrices for linear programming
+    c, A, b = initialize_LP_matrices(rpd, H)
+
+    # bounds on w are [-Inf, Inf] while bounds on slack are [0, Inf]
+    lb = vcat([-Inf, -Inf], zeros(nH+2))
+    ub = fill(Inf, nH+4)
+
+    # Set iterative parameters and iterate until converged
+    iter, dist = 0, 10.0
+    while (iter < maxiter) && (dist > tol)
+        # Compute the current worst values for each agent
+        _w1 = worst_value_1(rpd, H, C)
+        _w2 = worst_value_2(rpd, H, C)
```

Comment:
If `[-1, 0]` and `[0, -1]` are contained in `H`, which is the case if `nH` is a multiple of 4, are these values equal to the values of `C` that corresponding to `[-1, 0]` and `[0, -1]`?